C error when mutating a field after an `or { return }` block

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.4 86c0271.8215f21
OS: windows, Microsoft Windows 10 Pro v19045 64-bit
Processor: 20 cpus, 64bit, little endian, 

getwd: C:\Users\user\Desktop\Code\leaflang
vexe: C:\Users\user\Desktop\Code\v\v.exe
vexe mtime: 2024-02-13 10:54:11

vroot: OK, value: C:\Users\user\Desktop\Code\v
VMODULES: OK, value: C:\Users\user\.vmodules
VTMP: OK, value: C:\Users\user\AppData\Local\Temp\v_0

Git version: git version 2.42.0.windows.1
Git vroot status: weekly.2024.07-7-g8215f215
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,

operable program or batch file.


thirdparty/tcc status: thirdparty-windows-amd64 b99a453d

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg test.v`
```v
struct Tree {
mut:
	nr_elems int
	parent   ?&Tree
}

fn (mut t Tree) set_nr_elems(name string, value int) {
	t.parent or { return }.nr_elems = value
}

fn main() {
	parent := Tree{
		nr_elems: 11
	}
	mut child := Tree{
		parent: &parent
	}
	child.set_nr_elems('Buzz', 123)
}

```

**What did you expect to see?**

Program compiles


**What did you see instead?**
```
==================
C:/Users/user/AppData/Local/Temp/v_0/test.01HPH1ZHMZQ427C8E826MHBY6B.tmp.c:6872: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/user/AppData/Local/Temp/v_0/test.01HPH1ZHMZQ427C8E826MHBY6B.tmp.c:13089: error: pointer expected
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -3670,7 +3670,7 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	if is_as_cast {
 		g.write('(')
 	}
-	if node.or_block.kind != .absent && !g.is_assign_lhs && g.table.sym(node.typ).kind != .chan {
+	if node.or_block.kind != .absent && g.table.sym(node.typ).kind != .chan {
 		is_ptr := sym.kind in [.interface_, .sum_type]
 		stmt_str := g.go_before_last_stmt().trim_space()
 		styp := g.typ(g.unwrap_generic(node.typ))

--- a/vlib/v/tests/option_selector_assign_test.v
+++ b/vlib/v/tests/option_selector_assign_test.v
@@ -1,0 +1,20 @@
+struct Tree {
+mut:
+	nr_elems int
+	parent   ?&Tree
+}
+
+fn (mut t Tree) set_nr_elems(name string, value int) {
+	t.parent or { return }.nr_elems = value
+}
+
+fn test_main() {
+	parent := Tree{
+		nr_elems: 11
+	}
+	mut child := Tree{
+		parent: &parent
+	}
+	child.set_nr_elems('Buzz', 123)
+	assert child.parent or { return }.nr_elems == 123
+}

